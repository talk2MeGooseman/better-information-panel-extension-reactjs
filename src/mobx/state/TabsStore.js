import { observable, computed, action, makeObservable } from "mobx";
import TabModel from "../model/TabModel";
import { uuid } from "../../services/Utils";
import {
  queryPanelInformation
} from "../../services/Ebs";
import {
  DEFAULT_BODY_TEXT,
  ACTIVE_STEP_1,
  ACTIVE_STEP_2,
  ACTIVE_STEP_3,
  ACTIVE_STEP_4,
} from "../../services/constants";
import { path } from 'ramda';

export default class TabsStore {
  tabs = [];
  client = null;
  saveState = "";
  loadingState = "pending";
  activeStep = ACTIVE_STEP_1;
  videoOverlayHeight = '80vh';
  videoComponentVisibility = true;
  videoComponentTransparent = true;
  videoToggleImageUrl = '';
  videoToggleButtonPosition = '';
  tabIndex = 0;

  constructor() {
    makeObservable(this, {
      tabs: observable,
      client: observable,
      saveState: observable,
      loadingState: observable,
      activeStep: observable,
      videoOverlayHeight: observable,
      videoComponentVisibility: observable,
      videoComponentTransparent: observable,
      videoToggleImageUrl: observable,
      videoToggleButtonPosition: observable,
      tabIndex: observable,
      tabCount: computed,
      fetchTabs: action,
      updateTabs: action,
      saveTabs: action
    });
  }

  get tabCount() {
    return this.tabs.length;
  }

  setVideoComponentVisability(value) {
    this.saveState = '';
    this.videoComponentVisibility = value;
  }

  setVideoComponentTransparency(value) {
    this.saveState = '';
    this.videoComponentTransparent = value;
  }

  setVideoToggleImageUrl(value) {
    this.saveState = '';
    this.videoToggleImageUrl = value;
  }

  setVideoToggleButtonPosition(value) {
    this.saveState = '';
    this.videoToggleButtonPosition = value;
  }

  addTab() {
    this.saveState = '';

    if (this.activeStep === ACTIVE_STEP_1) {
      this.activeStep = ACTIVE_STEP_2;
    }

    this.tabs.push(TabModel.fromJS(this, {
      id: uuid(),
      title: 'Add Title',
      textColor: '#000000',
      bgColor: '#ffffff',
      body: DEFAULT_BODY_TEXT,
    }));
  }

  fetchTabs() {
    this.loadingState = "pending"
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", ({ data }) => {
        this.loadingState = "done"

        const result = path(['channel', 'bipConfig'], data)
        console.log(result)

        if (!result.tabs) {
          this.addTab();
          return;
        }

        this.activeStep = ACTIVE_STEP_4;

        // Check if the setting are defined from server, could be undefined
        // And screw shit up, use default state
        if (result.transparent !== undefined) {
          this.videoComponentTransparent = result.transparent;
        }
        if (result.visibility !== undefined) {
          this.videoComponentVisibility = result.visibility;
        }
        if (result.togglePosition !== undefined) {
          this.videoToggleButtonPosition = result.togglePosition;
        }

        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  updateTabs() {
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", result => {
        if (!result.tabs) {
          this.addTab();
          return;
        }
        this.activeStep = ACTIVE_STEP_4;
        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  saveTabs() {
    this.saveState = "pending"
      setPanelInformation(this.token, this.toJSON()).then(
        // inline created action
        action("fetchSuccess", result => {
          this.saveState = "done"
          this.activeStep = ACTIVE_STEP_4;
        }),
        // inline created action
        action("fetchError", error => {
          this.saveState = "error"
        })
      )
  }

  toJSON() {
    let jTabs = this.tabs.map((tab) => {
      return tab.toJSON();
    });

    return {
      tabs: jTabs,
      visibility: this.videoComponentVisibility,
      transparent: this.videoComponentTransparent,
      togglePosition: this.videoToggleButtonPosition,
    };
  }
}
