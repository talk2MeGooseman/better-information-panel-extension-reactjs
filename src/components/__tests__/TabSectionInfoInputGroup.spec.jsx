import React from "react";
import { createShallow, createMount} from '@material-ui/core/test-utils';

import TabSectionInfoInputGroup from "../TabSectionInfoInputGroup";
import TabsStore from "../../mobx/state/TabsStore";

describe('TabSectionInfoInputGroup', () => {
  let tabsStore, tab, shallow, mount;

  afterEach(() => {
    mount.cleanUp();
  });

  beforeEach(() => {
    shallow = createShallow({
      dive: true
    });
    mount = createMount();
    tabsStore = new TabsStore();
    tabsStore.addTab();
    tab = tabsStore.tabs[0];
  });

  it('matches default state snapshot', () => {
    const wrapper = shallow(<TabSectionInfoInputGroup index={1} tabData={tab} tabsStore={tabsStore} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('sets form values from tab settings', () => {
    tab.title = "Hey";
    tab.textColor = "#505050";
    tab.bgColor = "#444444";

    const wrapper = shallow(<TabSectionInfoInputGroup tabData={tab} tabsStore={tabsStore} />);
    expect(wrapper.find('TextField').prop('value')).toEqual('Hey');
    expect(wrapper.find('ColorPicker').at(0).prop('color')).toEqual('#505050');
    expect(wrapper.find('ColorPicker').at(1).prop('color')).toEqual('#444444');
  });

  it('title onChange updates store tab title', () => {
    tab.title = "Hey";

    const wrapper = shallow(<TabSectionInfoInputGroup tabData={tab} tabsStore={tabsStore} />);
    let field = wrapper.find('TextField').first()
    field.simulate('change', {
      target: {
        value: "Bubba Gump Shrimp"
      }
    });
    expect(tab.title).toEqual("Bubba Gump Shrimp");
  });

  it('delete should be disabled if only 1 tab', () => {
    const wrapper = mount(<TabSectionInfoInputGroup tabData={tab} tabsStore={tabsStore} />);
    expect(wrapper.find('CardHeader').first().prop('action')).toBeNull();
  });

  it('delete should be enabled if more then one tab', () => {
    tabsStore.addTab();
    const wrapper = mount(<TabSectionInfoInputGroup tabData={tab} tabsStore={tabsStore} />);
    expect(wrapper.find('CardHeader').first().prop('action')).not.toBeNull();
  });

  it('click delete should remove tab', () => {
    tabsStore.addTab();
    const wrapper = mount(<TabSectionInfoInputGroup tabData={tab} tabsStore={tabsStore} />);
    expect(tabsStore.tabCount).toEqual(2);
    wrapper.find('IconButton').simulate('click');
    expect(tabsStore.tabCount).toEqual(1);
  });
});