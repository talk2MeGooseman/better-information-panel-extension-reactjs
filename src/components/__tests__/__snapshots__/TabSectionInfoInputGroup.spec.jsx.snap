// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabSectionInfoInputGroup matches default state snapshot 1`] = `
<WithStyles(Card)
  className="_class-card-2"
>
  <WithStyles(CardHeader)
    action={null}
    title="Tab 1"
  />
  <WithStyles(CardContent)
    className="_class-cardContent-3"
  >
    <WithStyles(FormGroup)
      className="_class-inputField-1"
    >
      <TextField
        id="title-1"
        label="Tab Title"
        onChange={[Function]}
        required={false}
        select={false}
        value="Add Title"
      />
    </WithStyles(FormGroup)>
    <WithStyles(FormGroup)
      className="_class-inputField-1"
    >
      <WithStyles(Typography)
        variant="caption"
      >
        Text Color
      </WithStyles(Typography)>
      <ColorPicker
        color="#000000"
        handleChange={[Function]}
      />
    </WithStyles(FormGroup)>
    <WithStyles(FormGroup)
      className="_class-inputField-1"
    >
      <WithStyles(Typography)
        variant="caption"
      >
        Background Color
      </WithStyles(Typography)>
      <ColorPicker
        color="#ffffff"
        handleChange={[Function]}
      />
    </WithStyles(FormGroup)>
  </WithStyles(CardContent)>
</WithStyles(Card)>
`;
